WalletWasabi.Fluent/ViewModels/RecoverWalletPageViewModel.cs

```diff
@@ -0,0 +1,119 @@
+using NBitcoin;
+using ReactiveUI;
+using Splat;
+using WalletWasabi.Gui;
+using WalletWasabi.Gui.Validation;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Fluent.ViewModels
+{
+    public class RecoveryPageViewModel : NavBarItemViewModel
+    {
+        private readonly ObservableAsPropertyHelper<Mnemonic?> _currentMnemonic;
+        private readonly ObservableAsPropertyHelper<bool> _isMnemonicValid;
+        private ObservableCollection<string> _mnemonics;
+        private IEnumerable _suggestions;
+        private int _itemCountLimit;
+        private bool _restrictInputToSuggestions;
+        private object _selectedTag;
+
+        public ObservableCollection<string> Mnemonics
+        {
+            get => _mnemonics;
+            set => this.RaiseAndSetIfChanged(ref _mnemonics, value);
+        }
+
+        public IEnumerable Suggestions
+        {
+            get => _suggestions;
+            set => this.RaiseAndSetIfChanged(ref _suggestions, value);
+        }
+
+        public int ItemCountLimit
+        {
+            get => _itemCountLimit;
+            set => this.RaiseAndSetIfChanged(ref _itemCountLimit, value);
+        }
+
+        public bool RestrictInputToSuggestions
+        {
+            get => _restrictInputToSuggestions;
+            set => this.RaiseAndSetIfChanged(ref _restrictInputToSuggestions, value);
+        }
+
+        public object SelectedTag
+        {
+            get => _selectedTag;
+            set => this.RaiseAndSetIfChanged(ref _selectedTag, value);
+        }
+
+        public RecoveryPageViewModel(IScreen screen) : base(screen)
+        {
+            Global = Locator.Current.GetService<Global>();
+
+            Title = "Recovery";
+
+            Suggestions = new Mnemonic(Wordlist.English, WordCount.Twelve).WordList.GetWords();
+            RestrictInputToSuggestions = true;
+            ItemCountLimit = (int) WordCount.Twelve;
+
+            Mnemonics = new ObservableCollection<string>();
+
+            _currentMnemonic = Mnemonics.ToObservableChangeSet().ToCollection()
+                .Select(x => x.Count == 12 ? new Mnemonic(GetTagsAsConcatString()) : default)
+                .ToProperty(this, x => x.CurrentMnemonics);
+
+            _isMnemonicValid = this.WhenAnyValue(x => x.CurrentMnemonics)
+                .Select(x => x?.WordList?.WordCount == 12 && (x?.IsValidChecksum ?? false))
+                .ToProperty(this, x => x.IsMnemonicValid);
+
+            this.WhenAnyValue(x => x.SelectedTag)
+                .Where(x => x is string s && !string.IsNullOrEmpty(s))
+                .Select(x => x as string)
+                .Subscribe(AddTag!);
+
+            this.ValidateProperty(x => x.Mnemonics, ValidateMnemonics);
+
+            // A hack for validations system...
+            Mnemonics.CollectionChanged += MnemonicsChanged;
+        }
+
+        private void AddTag(string tagString)
+        {
+            Mnemonics.Add(tagString);
+            SelectedTag = null!;
```

Comment:
no bangs! sorry... if your intentionally setting SelectedTag to nulll then make the type have ?.